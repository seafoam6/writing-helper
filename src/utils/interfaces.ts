export interface ITodo {
  id: number;
  description: string;
  created_on: string;
  active: boolean;
}
export interface ITodosState {
  todos: {
    ids: number[];
    byId: {
      [Key: number]: ITodo;
    };
    active: number[];
    inactive: number[];
  };
}

export interface DatabaseResponseRow {
  description: string;
  created_on: string;
  id: number;
  active: boolean;
}

//Todo: change this to a better name
export interface CreateTodoResponse {
  data: {
    rows: DatabaseResponseRow[];
  };
}

export interface IContextProps {
  state: ITodosState;
  dispatch: ({ type }: { type: string }) => void;
}
