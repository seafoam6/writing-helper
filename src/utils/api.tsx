import axios from 'axios';
import { AxiosResponse } from 'axios';

const todosUrl = 'http://localhost:3000/todos';

const fetchTodos = () => {
  return axios
    .get(todosUrl)
    .then((response: void | AxiosResponse) => {
      return response;
    })
    .catch(error => {
      // handle error
      console.log(error);
    });
};

export { fetchTodos };
