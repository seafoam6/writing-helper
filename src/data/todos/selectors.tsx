import { ITodosState } from '../../utils/interfaces';
import { createSelector } from 'reselect';

export const getTodos = (state: ITodosState) => state.todos;

export const getTodoIds = (state: ITodosState) => state.todos.ids;

export const getAllTodos = createSelector(
  getTodos,
  getTodoIds,
  (todos, ids) => {
    return ids.map(id => todos.byId[id]);
  }
);

export const getTodoById = (state: ITodosState, id: string) =>
  state.todos.byId[id];
