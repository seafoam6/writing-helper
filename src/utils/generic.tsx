import { ITodo } from '../utils/interfaces';

export const arrOfNumbers = (count: number): number[] => {
  let i = 0;
  const thing: number[] = [];
  while (i < count) {
    i++;
    thing.push(i);
  }
  return thing;
};

export const getRandomTodo = (list: ITodo[]) => {
  const active = list.filter(i => i.active);
  return active[Math.floor(Math.random() * list.length)];
};
