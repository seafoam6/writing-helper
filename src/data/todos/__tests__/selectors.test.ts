import { ITodosState } from '../../../utils/interfaces';
import * as selectors from '../selectors';

const state: ITodosState = {
  todos: {
    ids: [18, 19, 20] as number[],
    byId: {
      '18': {
        id: 18,
        description: 'ðŸ‘€ðŸŽ¬read scenes, leave notes',
        created_on: '2019-06-02T18:06:42.051Z',
        active: true
      },
      '19': {
        id: 19,
        description: 'ðŸ‘€ðŸŽ¬read scenes, rewrite outline to match notes',
        created_on: '2019-06-02T18:06:42.111Z',
        active: true
      },
      '20': {
        id: 20,
        description: 'ðŸŽ¬ðŸ—’Split acts into sub acts',
        created_on: '2019-06-02T18:06:42.128Z',
        active: false
      }
    },
    active: [18, 19] as number[],
    inactive: [20] as number[]
  }
};

describe('getTodos', () => {
  it('returns todos piece of state', () => {
    expect(selectors.getTodos(state)).toEqual(state.todos);
  });
});

describe('getTodoIds', () => {
  it('returns ids', () => {
    expect(selectors.getTodoIds(state)).toEqual([18, 19, 20]);
  });
});

describe('getAllTodos', () => {
  it('returns all todos', () => {
    expect(selectors.getAllTodos(state)).toMatchSnapshot();
  });
});

describe('getTodoById', () => {
  it('returns all todos', () => {
    expect(selectors.getTodoById(state, 18)).toMatchSnapshot();
  });
});
