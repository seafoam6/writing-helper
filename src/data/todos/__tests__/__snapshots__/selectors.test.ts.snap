// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getAllTodos returns all todos 1`] = `
Array [
  Object {
    "active": true,
    "created_on": "2019-06-02T18:06:42.051Z",
    "description": "ðŸ‘€ðŸŽ¬read scenes, leave notes",
    "id": 18,
  },
  Object {
    "active": true,
    "created_on": "2019-06-02T18:06:42.111Z",
    "description": "ðŸ‘€ðŸŽ¬read scenes, rewrite outline to match notes",
    "id": 19,
  },
  Object {
    "active": false,
    "created_on": "2019-06-02T18:06:42.128Z",
    "description": "ðŸŽ¬ðŸ—’Split acts into sub acts",
    "id": 20,
  },
]
`;

exports[`getTodoById returns all todos 1`] = `
Object {
  "active": true,
  "created_on": "2019-06-02T18:06:42.051Z",
  "description": "ðŸ‘€ðŸŽ¬read scenes, leave notes",
  "id": 18,
}
`;
